docs/csharp/programming-guide/events/codesnippet/CSharp/how-to-use-a-dictionary-to-store-event-instances_1.cs

```diff
@@ -2,33 +2,44 @@
     using System.Collections.Generic;    
 
     public delegate void EventHandler1(int i);
+
     public delegate void EventHandler2(string s);
 
     public class PropertyEventsSample
     {
-        private Dictionary<string, Delegate> eventTable;
-
+        private readonly Dictionary<string, Delegate> _eventTable;
+        private readonly List<EventHandler1> _event1List = new List<EventHandler1>();
+        private readonly List<EventHandler2> _event2List = new List<EventHandler2>();
         public PropertyEventsSample()
         {
-            eventTable = new Dictionary<string, Delegate>();
-            eventTable.Add("Event1", null);
-            eventTable.Add("Event2", null);
+            _eventTable = new Dictionary<string, Delegate>
+            {
+                {"Event1", null},
+                {"Event2", null}
+            };
         }
 
         public event EventHandler1 Event1
         {
             add
             {
-                lock (eventTable)
+                _event1List.Add(value);
```

Comment:
One more remark: by the same reasoning as access to `_eventTable` is synchronized with the `lock` statement, shouldn't the access to `_event1List` (and `_event2List`) be synchronized as well?

Reply:
Also correct. This makes it more complicated because it will have to acquire multiple structures in a single locked region.

Reply:
@BillWagner can we go here with only one lock object? I mean: move the `lock (_eventTable)` line one line up?

Reply:
Yes.

I think the lock on Line 36 (in the corresponding remove) should also move up one line to cover the `eventList.Remove()` call.

Also, a similar change to the `EventHandler2` `add` and `remove` accessors. Do you agree?

Reply:
Yes. And not to confuse readers, I'll just introduce a dedicated `handlerStorageLock` object.